src/prng/isaac64.rs

```diff
@@ -181,45 +193,70 @@ impl Isaac64Rng {
 
         self.a = a;
         self.b = b;
-        self.cnt = RAND_SIZE as u32;
+        self.index = 1;
     }
 }
 
 impl Rng for Isaac64Rng {
     #[inline]
     fn next_u32(&mut self) -> u32 {
-        self.next_u64() as u32
+        // Using a local variable for `index`, and checking the size avoids a
+        // bounds check later on.
+        let mut index = self.index as usize - 1;
+        if index >= RAND_SIZE * 2 {
+            self.isaac64();
+            index = 0;
+        }
+
+        let value;
+        if cfg!(target_endian = "little") {
+            // Index as if this is a u32 slice.
+            let rsl = unsafe { &*(&mut self.rsl as *mut [u64; RAND_SIZE]
+                                                as *mut [u32; RAND_SIZE * 2]) };
+            value = rsl[index];
+        } else {
+            // Index into the u64 slice, rotate and truncate the result.
+            // Works always, also on big-endian systems, but is slower.
+            let tmp = self.rsl[index >> 1];
+            value = tmp as u32;
+            self.rsl[index >> 1] = tmp.rotate_right(32);
```

Comment:
Why not `tmp >> 32`? We've already used the other bits and won't reuse them (I hope)!